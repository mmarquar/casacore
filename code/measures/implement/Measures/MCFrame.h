//# MCFrame.h: Measure frame calculations proxy
//# Copyright (C) 1996,1997
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#if !defined(AIPS_MCFRAME_H)
#define AIPS_MCFRAME_H

#if defined(_AIX)
#pragma implementation ("MCFrame.cc")
#endif

//# Includes
#include <aips/aips.h>
#include <aips/Arrays/Vector.h>
#include <aips/Measures/Measure.h>
#include <aips/Measures/MeasFrame.h>

//# Forward Declarations
class MVDirection;
#if defined(AIPS_STDLIB)
#include <iosfwd.h>
#else
class ostream;
#endif

// <summary>
// Measure frame calculations proxy
// </summary>

// <use visibility=local>

// <reviewed reviewer="" date="yyyy/mm/dd" tests="tMeasure" demos="">
// </reviewed>

// <prerequisite>
// <li> <linkto class=Measure>Measure</linkto> class
// <li> <linkto class=MeasFrame>MeasFrame</linkto> class
// </prerequisite>
//
// <etymology>
// From Measure and Frame
// </etymology>
//
// <synopsis>
// The <linkto class=MeasFrame>MeasFrame</linkto> class contains the 'when
// and where' of an observed Measure. Calculations to get the appropiate
// value (e.g. the Earth's longitude) from this frame for conversions are
// done in this class, together with all the caching of (intermediate) results
// that can speed-up calculations.<br>
// The MCFrame class is used by the individual measure conversion classes
// (see <linkto class=MCBase>MCBase</linkto> class).<br>
// </synopsis>
//
// <example>
// <srcblock>
//	MEpoch my_epoch(Quantity(MeasData::MJDB1950,"d")); // an epoch
//	MCFrame frame(my_epoch);	// used in a frame
//	frame.set(obser);		// add observatory (an MPosition)
//	MEpoch::Convert conv(my_epoch, MEPoch::Ref(MEpoch::LAST, frame));
// </srcblock>
// The <em>conv</em> conversion engine will (transpararently) use the MCFrame
// class in calls from MCEpoch (the time conversions), which will be called
// by the MEpoch::Convert () operator.
// </example>
//
// <motivation>
// To separate the frame calculations from the Measure containers, to enable
// e.g. Tables to have Measures.
// </motivation>
//
// <todo asof="1997/04/17">
// </todo>

class MCFrame {

public:
  
  //# Friends
  // Delete a conversion frame data
  friend void MCFrameDelete(void *dmf);
  
  //# Constructors
  
  //# Operators
  
  //# General member functions
  // Attach an instant of MCFrame to belonging MeasFrame
  static void make(MeasFrame &in);
  // Get TDB in days
  Bool getTDB(Double &tdb);
  // Get the longitude (in rad)
  Bool getLong(Double &tdb);
  // Get the latitude (in rad)
  Bool getLat(Double &tdb);
  // Get the gecentric position (in m)
  Bool getRadius(Double &tdb);
  // Get the LAST (in days)
  Bool getLAST(Double &tdb);
  // Get the LAST (in rad)
  Bool getLASTr(Double &tdb);
  // Get J2000 coordinates (direction cosines)
  Bool getJ2000(MVDirection &tdb);
  // Get B1950 coordinates (direction cosines)
  Bool getB1950(MVDirection &tdb);
  // Get apparent coordinates (direction cosines)
  Bool getApp(MVDirection &tdb);
  // Get LSR radial velocity (m/s)
  Bool getLSR(Double &tdb);
  
private:
  //# Data
  // The belonging frame pointer
  MeasFrame myf;
  // The actual measure conversion values
  // <group>
  // Conversion to TDB time (due to some (for me) unsolvable dependency errors
  // not the proper MeasConvert* here)
  void *epConvTDB;
  // TDB time
  Double *epTDBp;
  // Conversion to LAST time
  void *epConvLAST;
  // LAST time
  Double *epLASTp;
  // Conversion to astronomical longitude/latitude
  void *posConvLong;
  // Longitude
  Vector<Double> *posLongp;
  // Conversion to J2000
  void *dirConvJ2000;
  // J2000 coordinates
  MVDirection *dirJ2000p;
  // Conversion to B1950
  void *dirConvB1950;
  // B1950 coordinates
  MVDirection *dirB1950p;
  // Conversion to apparent coordinates
  void *dirConvApp;
  // Apparent coordinates
  MVDirection *dirAppp;
  // Conversion to LSR radial velocity
  void *radConvLSR;
  // Radial velocity
  Double *radLSRp;
  // </group>
  
  //# Member functions
  // Constructor
  MCFrame(MeasFrame &inf);
  // Default constructor (not implemented)
  MCFrame();
  // Copy constructor (not implemented)
  MCFrame(const MCFrame &other);
  // Copy assignment (not implemented)
  MCFrame &operator=(const MCFrame &other);
  // Destructor
  ~MCFrame();
  // Create an instance of the MCFrame class
  void create();
  // Reset Epoch value
  void resetEpoch();
  // Reset Position value
  void resetPosition();
  // Reset Direction value
  void resetDirection();
  // Reset RadialVelocity value
  void resetRadialVelocity();
  // Make full Epoch
  void makeEpoch();
  // Make full Position
  void makePosition();
  // Make full Direction
  void makeDirection();
  // Make full RadialVelocity
  void makeRadialVelocity();
};

//# Global functions
// <summary> Global functions </summary>
// <group name=MeasFrameInterface>
// Delete a frame
void MCFrameDelete(void *dmf);
// Get double value for MeasFrame
Bool MCFrameGetdbl(void *dmf, uInt tp, Double &result);
// Get MVDirection value for MeasFrame
Bool MCFrameGetmvdir(void *dmf, uInt tp, MVDirection &result);
// </group>

#endif
