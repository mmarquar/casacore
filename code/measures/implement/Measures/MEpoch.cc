//# MEpoch.cc: A Measure: instant in time
//# Copyright (C) 1995,1996,1997
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

//# Includes
#ifdef __GNUG__
#include <aips/Measures/Quantum.h>
typedef Quantum<Double> gpp_mepoch_bug1;
#endif
#include <aips/Exceptions/Error.h>
#include <aips/Utilities/Assert.h>
#include <aips/RTTI/Register.h>
#include <aips/Measures/MEpoch.h>

//# Constructors
MEpoch::MEpoch() : 
  MeasBase<MVEpoch,MEpoch::Ref>() {}

MEpoch::MEpoch(const MVEpoch &dt) : 
  MeasBase<MVEpoch,MEpoch::Ref>(dt,MEpoch::DEFAULT) {}

MEpoch::MEpoch(const MVEpoch &dt, const MEpoch::Ref &rf) : 
  MeasBase<MVEpoch,MEpoch::Ref>(dt,rf) {}

MEpoch::MEpoch(const MVEpoch &dt, uInt rf) : 
  MeasBase<MVEpoch,MEpoch::Ref>(dt,rf) {}

MEpoch::MEpoch(const Quantity &dt) : 
  MeasBase<MVEpoch,MEpoch::Ref>(dt,MEpoch::DEFAULT) {}

MEpoch::MEpoch(const Quantity &dt, const MEpoch::Ref &rf) : 
  MeasBase<MVEpoch,MEpoch::Ref>(dt,rf) {}

MEpoch::MEpoch(const Quantity &dt, uInt rf) : 
  MeasBase<MVEpoch,MEpoch::Ref>(dt,rf) {}

MEpoch::MEpoch(const Measure *dt) :
  MeasBase<MVEpoch,MEpoch::Ref>(dt) {}

MEpoch::MEpoch(const MeasValue *dt) :
  MeasBase<MVEpoch,MEpoch::Ref>(*(MVEpoch*)dt,
				MEpoch::DEFAULT) {}

//# Destructor
MEpoch::~MEpoch() {}

//# Operators

//# Member functions

const String &MEpoch::tellMe() const {
  return MEpoch::showMe();
}

const String &MEpoch::showMe() {
  static const String name("Epoch");
  return name;
}

uInt MEpoch::type() const {
  return Register((MEpoch *)0);
}

void MEpoch::assert(const Measure &in) {
  if (in.type() != Register((MEpoch *)0)) {
    throw(AipsError("Illegal Measure type argument: " +
		    MEpoch::showMe()));
  };
}

const String &MEpoch::showType(uInt tp) {
  static const String tname[MEpoch::N_Types] = {
    "LAST",
    "LMST",
    "GMST1",
    "GAST",
    "UT1",
    "UT2",
    "UTC",
    "TAI",
    "TDT",
    "TCG",
    "TDB",
    "TCB"};
  DebugAssert((tp & ~MEpoch::EXTRA) < MEpoch::N_Types, AipsError);
  return tname[tp & ~MEpoch::EXTRA];
}

Bool MEpoch::giveMe(const String &in, MEpoch::Ref &mr) {
  static const Int N_name = 17;
  static const String tname[N_name] = {
    "LAST",
    "LMST",
    "GMST1",
    "GAST",
    "UT1",
    "UT2",
    "UTC",
    "TAI",
    "TDT",
    "TCG",
    "TDB",
    "TCB",
    "IAT",
    "GMST",
    "TT",
    "ET",
    "UT"};
  
  static const uInt oname[N_name] = {
    MEpoch::LAST,
    MEpoch::LMST,
    MEpoch::GMST1,
    MEpoch::GAST,
    MEpoch::UT1,
    MEpoch::UT2,
    MEpoch::UTC,
    MEpoch::TAI,
    MEpoch::TDT,
    MEpoch::TCG,
    MEpoch::TDB,
    MEpoch::TCB,
    MEpoch::TAI,
    MEpoch::GMST1,
    MEpoch::TDT,
    MEpoch::TDT,
    MEpoch::UT1};
  
  uInt i = Measure::giveMe(in, N_name, tname);
  if (i>=N_name) {
    mr = MEpoch::Ref();
    return False;
  } else {
    mr = MEpoch::Ref(oname[i]);
  };
  return True;
}

Quantity MEpoch::get(const Unit &inunit) const {
  return (data.getTime().get(inunit));
}

Measure *MEpoch::clone() const {
  return (new MEpoch(*this));
}
