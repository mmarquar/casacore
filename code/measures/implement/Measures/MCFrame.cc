//# MCFrame.cc: Measure frame calculations proxy
//# Copyright (C) 1996, 1997, 1998, 1999
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

//# Includes
#include <aips/Exceptions/Error.h>
#include <aips/Mathematics/Math.h>
#include <aips/Mathematics/Constants.h>
#include <aips/Quanta/MVAngle.h>
#include <aips/Measures/MEpoch.h>
#include <aips/Measures/MCEpoch.h>
#include <aips/Measures/MPosition.h>
#include <aips/Measures/MCPosition.h>
#include <aips/Measures/MDirection.h>
#include <aips/Measures/MCDirection.h>
#include <aips/Measures/MRadialVelocity.h>
#include <aips/Measures/MCRadialVelocity.h>
#include <aips/Measures/MCFrame.h>
#include <aips/Measures/MeasConvert.h>

// MCFrame class
 
//# Constructors
MCFrame::MCFrame(MeasFrame &inf) :
  myf(inf),
  epConvTDB(0), epTDBp(0), 
  epConvLAST(0), epLASTp(0), 
  posConvLong(0), posLongp(0), posITRFp(0),
  dirConvJ2000(0), dirJ2000p(0),
  dirConvB1950(0), dirB1950p(0),
  dirConvApp(0), dirAppp(0),
  radConvLSR(0), radLSRp(0) {
    myf.setMCFramePoint((void *) this);
    myf.setMCFrameDelete(MCFrameDelete);
    myf.setMCFrameGetdbl(MCFrameGetdbl);
    myf.setMCFrameGetmvdir(MCFrameGetmvdir);
    myf.setMCFrameGetmvpos(MCFrameGetmvpos);
    create();
    myf.unlock();
}

// Destructor
MCFrame::~MCFrame() {
  delete (MEpoch::Convert *) epConvTDB;
  delete epTDBp;
  delete (MEpoch::Convert *) epConvLAST;
  delete epLASTp;
  delete (MPosition::Convert *) posConvLong;
  delete posLongp;
  delete posITRFp;
  delete (MDirection::Convert *) dirConvJ2000;
  delete dirJ2000p;
  delete (MDirection::Convert *) dirConvB1950;
  delete dirB1950p;
  delete (MDirection::Convert *) dirConvApp;
  delete dirAppp;
  delete (MRadialVelocity::Convert *) radConvLSR;
  delete radLSRp;
}

// Operators

// General member functions

void MCFrame::make(MeasFrame &in) {
  if (!in.empty() && !in.getMCFramePoint()) {
    MCFrame *tmp = new MCFrame(in);
    if (!tmp) {};		// to stop compiler warnings
  };
  if (!in.empty()) {
    ((MCFrame *)(in.getMCFramePoint()))->create();
  };
}

void MCFrame::resetEpoch() {
  if (epTDBp) {
    delete epTDBp; epTDBp = 0;
  };
  if (epLASTp) {
    delete epLASTp; epLASTp = 0;
  };
  if (dirAppp) {
    delete dirAppp; dirAppp = 0;
  };
  if (radLSRp) {
    delete radLSRp; radLSRp = 0;
  };
}

void MCFrame::resetPosition() {
  if (posLongp) {
    delete posLongp; posLongp = 0;
    delete posITRFp; posITRFp = 0;
  };
  if (epLASTp) {
    delete epLASTp; epLASTp = 0;
  };
}

void MCFrame::resetDirection() {
  if (dirJ2000p) {
    delete dirJ2000p; dirJ2000p = 0;
  };
  if (dirB1950p) {
    delete dirB1950p; dirB1950p = 0;
  };
  if (dirAppp) {
    delete dirAppp; dirAppp = 0;
  };
  if (radLSRp) {
    delete radLSRp; radLSRp = 0;
  };
}

void MCFrame::resetRadialVelocity() {
  if (radLSRp) {
    delete radLSRp; radLSRp = 0;
  };
}

Bool MCFrame::getTDB(Double &tdb) {
  if (myf.epoch()) {
    if (!epTDBp) {
      epTDBp = new Double;
      *(epTDBp) = (*(MEpoch::Convert *)
		   (epConvTDB))().getValue().get();
    };
    tdb = *(epTDBp);
    return True;
  };
  tdb = 0;
  return False;
}

Bool MCFrame::getLong(Double &tdb) {
  if (myf.position()) {
    if (!posLongp) {
      posLongp = new Vector<Double>(3);
      posITRFp = new MVPosition;
      *(posITRFp) = (*(MPosition::Convert *)
		     (posConvLong))().getValue();
      *(posLongp) = posITRFp->get();
    };
    tdb = MVAngle(posLongp->operator()(1))(-0.5);
    return True;
  };
  tdb = 0;
  return False;
}

Bool MCFrame::getLat(Double &tdb) {
  if (myf.position()) {
    if (!posLongp) {
      posLongp = new Vector<Double>(3);
      posITRFp = new MVPosition;
      *(posITRFp) = (*(MPosition::Convert *)
		     (posConvLong))().getValue();
      *(posLongp) = posITRFp->get();
    };
    tdb = posLongp->operator()(2);
    return True;
  };
  tdb = 0;
  return False;
}

Bool MCFrame::getITRF(MVPosition &tdb) {
  if (myf.position()) {
    if (!posLongp) {
      posLongp = new Vector<Double>(3);
      posITRFp = new MVPosition;
      *(posITRFp) = (*(MPosition::Convert *)
		     (posConvLong))().getValue();
      *(posLongp) = posITRFp->get();
    };
    tdb = *posITRFp;
    return True;
  };
  tdb = (Double) 0.0;
  return False;
}

Bool MCFrame::getRadius(Double &tdb) {
  if (myf.position()) {
    if (!posLongp) {
      posLongp = new Vector<Double>(3);
      posITRFp = new MVPosition;
      *(posITRFp) = (*(MPosition::Convert *)
		     (posConvLong))().getValue();
      *(posLongp) = posITRFp->get();
    };
    tdb = posLongp->operator()(0);
    return True;
  };
  tdb = 0;
  return False;
}

Bool MCFrame::getLAST(Double &tdb) {
  if (myf.epoch()) {
    if (!epLASTp) {
      epLASTp = new Double;
      *(epLASTp) = (*(MEpoch::Convert *)
		    (epConvLAST))().getValue().get();
    };
    tdb = fmod(*(epLASTp),1.0);
    return True;
  };
  tdb = 0;
  return False;
}

Bool MCFrame::getLASTr(Double &tdb) {
  Bool tmp = MCFrame::getLAST(tdb);
  tdb *= C::circle;
  return tmp;
}

Bool MCFrame::getJ2000(MVDirection &tdb) {
  if (myf.direction()) {
    if (!dirJ2000p) {
      dirJ2000p = new MVDirection;
      *(dirJ2000p) = (*(MDirection::Convert *)
		      (dirConvJ2000))().getValue();
    };
    tdb = *(dirJ2000p);
    return True;
  };
  tdb = (Double) 0.0;
  return False;
}

Bool MCFrame::getB1950(MVDirection &tdb) {
  if (myf.direction()) {
    if (!dirB1950p) {
      dirB1950p = new MVDirection;
      *(dirB1950p) = (*(MDirection::Convert *)
		      (dirConvB1950))().getValue();
    };
    tdb = *(dirB1950p);
    return True;
  };
  tdb = (Double) 0.0;
  return False;
}

Bool MCFrame::getApp(MVDirection &tdb) {
  if (myf.direction()) {
    if (!dirAppp) {
      dirAppp = new MVDirection;
      *(dirAppp) = (*(MDirection::Convert *)
		    (dirConvApp))().getValue();
    };
    tdb = *(dirAppp);
    return True;
  };
  tdb = (Double) 0.0;
  return False;
}

Bool MCFrame::getLSR(Double &tdb) {
  if (myf.radialVelocity()) {
    if (!radLSRp) {
      radLSRp = new Double;
      *(radLSRp) = (*(MRadialVelocity::Convert *)
		    (radConvLSR))().getValue();
    };
    tdb = *(radLSRp);
    return True;
  };
  tdb = (Double) 0.0;
  return False;
}

void MCFrame::create() {
  if (myf.getEpset()) {
    makeEpoch();
    myf.setEpset(False);
    myf.setEpreset(False);
  };
  if (myf.getPosset()) {
    makePosition();
    myf.setPosset(False);
    myf.setPosreset(False);
  };
  if (myf.getDirset()) {
    makeDirection();
    myf.setDirset(False);
    myf.setDirreset(False);
  };
  if (myf.getRadset()) {
    makeRadialVelocity();
    myf.setRadset(False);
    myf.setRadreset(False);
  };
  if (myf.getEpreset()) {
    resetEpoch();
    myf.setEpreset(False);
  };
  if (myf.getPosreset()) {
    resetPosition();
    myf.setPosreset(False);
  };
  if (myf.getDirreset()) {
    resetDirection();
    myf.setDirreset(False);
  };
  if (myf.getRadreset()) {
    resetRadialVelocity();
    myf.setRadreset(False);
  };
}

void MCFrame::makeEpoch() {
  static const MEpoch::Ref REFTDB 
    = MEpoch::Ref(MEpoch::TDB);
  delete (MEpoch::Convert *) epConvTDB;
  epConvTDB = new MEpoch::Convert(*(myf.epoch()),
				  REFTDB);
  if (epTDBp) {
    delete epTDBp; epTDBp = 0;
  };
  if (epConvLAST) {
    myf.lock();
    delete (MEpoch::Convert *) epConvLAST;
    epConvLAST = 0;
  };
  epConvLAST = new MEpoch::Convert(*(myf.epoch()),
				   MEpoch::Ref(MEpoch::LAST, this->myf));
  if (epConvLAST) {
    myf.unlock();
  };
  if (epLASTp) {
    delete epLASTp; epLASTp = 0;
  };
  if (dirAppp) {
    delete dirAppp; dirAppp = 0;
  };
  if (radLSRp) {
    delete radLSRp; radLSRp = 0;
  };
}

void MCFrame::makePosition() {
  static const MPosition::Ref REFLONG 
    = MPosition::Ref(MPosition::ITRF);
  delete (MPosition::Convert *) posConvLong;
  posConvLong = new MPosition::Convert(*(myf.position()),
				       REFLONG);
  if (posLongp) {
    delete posLongp; posLongp = 0;
    delete posITRFp; posITRFp = 0;
  };
  if (epLASTp) {
    delete epLASTp; epLASTp = 0;
  };
  if (radLSRp) {
    delete radLSRp; radLSRp = 0;
  };
}

void MCFrame::makeDirection() {
  static const MDirection::Ref REFJ2000
    = MDirection::Ref(MDirection::J2000);
  if (dirConvJ2000) {
    myf.lock();
    delete (MDirection::Convert *) dirConvJ2000;
    dirConvJ2000 = 0;
  };
  dirConvJ2000 = new MDirection::Convert(*(myf.direction()),
					 MDirection::Ref(MDirection::J2000,
							 this->myf));
  if (dirConvJ2000) myf.unlock();

  static const MDirection::Ref REFB1950
    = MDirection::Ref(MDirection::B1950);
  if (dirConvB1950) {
    myf.lock();
    delete (MDirection::Convert *) dirConvB1950;
    dirConvB1950 = 0;
  };
  dirConvB1950 = new MDirection::Convert(*(myf.direction()),
					 MDirection::Ref(MDirection::B1950,
							 this->myf));
  if (dirConvB1950) myf.unlock();

  if (dirConvApp) {
    myf.lock();
    delete (MDirection::Convert *) dirConvApp;
    dirConvApp = 0;
  };
  dirConvApp = new MDirection::Convert(*(myf.direction()),
				       MDirection::Ref(MDirection::APP,
						       this->myf));
  if (dirConvApp) {
    myf.unlock();
  };
  if (dirJ2000p) {
    delete dirJ2000p; dirJ2000p = 0;
  };
  if (dirB1950p) {
    delete dirB1950p; dirB1950p = 0;
  };
  if (dirAppp) {
    delete dirAppp; dirAppp = 0;
  };
  if (radLSRp) {
    delete radLSRp; radLSRp = 0;
  };
}

void MCFrame::makeRadialVelocity() {
  static const MRadialVelocity::Ref REFLSR 
    = MRadialVelocity::Ref(MRadialVelocity::LSR);
  delete (MRadialVelocity::Convert *) radConvLSR;
  radConvLSR = new MRadialVelocity::Convert(*(myf.radialVelocity()),
					    REFLSR);
  if (radLSRp) {
    delete radLSRp; radLSRp = 0;
  };
}

void MCFrameDelete(void *dmf) {
  delete (MCFrame *) dmf;
}

Bool MCFrameGetdbl(void *dmf, uInt tp, Double &result) {
  try {
    switch (tp) {
      
    case MeasFrame::GetTDB:
      return ((MCFrame *) dmf)->getTDB(result);
      break;
      
    case MeasFrame::GetLong:
      return ((MCFrame *) dmf)->getLong(result);
      break;
      
    case MeasFrame::GetLat:
      return ((MCFrame *) dmf)->getLat(result);
      break;
      
    case MeasFrame::GetRadius:
      return ((MCFrame *) dmf)->getRadius(result);
      break;
      
    case MeasFrame::GetLAST:
      return ((MCFrame *) dmf)->getLAST(result);
      break;
      
    case MeasFrame::GetLASTr:
      return ((MCFrame *) dmf)->getLASTr(result);
      break;
      
    case MeasFrame::GetLSR:
      return ((MCFrame *) dmf)->getLSR(result);
      break;
      
    default:
      break;
    };
  } catch (AipsError x) {
  } end_try;
  
  result = 0;
  return False;
}

Bool MCFrameGetmvdir(void *dmf, uInt tp, MVDirection &result) {
  try {
    switch (tp) {
      
    case MeasFrame::GetJ2000:
      return ((MCFrame *) dmf)->getJ2000(result);
      break;
      
    case MeasFrame::GetB1950:
      return ((MCFrame *) dmf)->getB1950(result);
      break;
      
    case MeasFrame::GetApp:
      return ((MCFrame *) dmf)->getApp(result);
      break;
      
    default:
      break;
    };
  } catch (AipsError x) {
  } end_try;
  MVDirection tmp;
  result = tmp;
  return False;
}

Bool MCFrameGetmvpos(void *dmf, uInt tp, MVPosition &result) {
  try {
    switch (tp) {
       
    case MeasFrame::GetITRF:
      return ((MCFrame *) dmf)->getITRF(result);
      break;
      
    default:
      break;
    };
  } catch (AipsError x) {
  } end_try;
  MVPosition tmp;
  result = tmp;
  return False;
}
