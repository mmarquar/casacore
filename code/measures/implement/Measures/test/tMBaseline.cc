//# tBaseline.cc: This program test Baseline and UVW functions
//# Copyright (C) 1998
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This program is free software; you can redistribute it and/or modify it
//# under the terms of the GNU General Public License as published by the Free
//# Software Foundation; either version 2 of the License, or (at your option)
//# any later version.
//#
//# This program is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
//# more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with this program; if not, write to the Free Software Foundation, Inc.,
//# 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

//# Includes
#include <aips/aips.h>
#include <aips/Exceptions/Error.h>
#include <aips/Measures.h>
#include <aips/Measures/Aberration.h>
#include <aips/Measures/MEpoch.h>
#include <aips/Measures/MDirection.h>
#include <aips/Measures/MPosition.h>
#include <trial/Measures/MBaseline.h>
#include <aips/Measures/MCFrame.h>
#include <aips/Arrays/ArrayMath.h>
#include <aips/Arrays/ArrayLogical.h>

main()
{
    try {
      	cout << "Test measure class MBaseline" << endl;
	cout << "--------------------------------------" << endl;


	MEpoch tbm(Quantity(50927.92931, "d"));
	MPosition pos(MVPosition(-4750915.84032, 2792906.17778, 
				 -3200483.75028), 
		      MPosition::ITRF);
	MeasFrame mf(tbm, pos);
	MVBaseline mvb0(100 ,10, 0);
	cout << "Baseline: " << mvb0 << endl;

	MBaseline::Ref mbref0(MBaseline::ITRF, mf);
	MBaseline mb0(mvb0, mbref0);

	cout << "Baseline: " << mb0 << endl;
	cout << "Baseline reference: " << mbref0 << endl;

	MBaseline::Ref mbref1(MBaseline::J2000);

	cout << "Baseline reference: " << mbref1 << endl;

	cout << "Test Baseline conversion ..." << endl;
	cout << "--------------------------------------" << endl;
	MBaseline::Convert bconv(mb0, mbref1);
	cout << "Converted " << mb0 << endl <<
	  " to " << mbref1 << endl <<
	  " as " << bconv() << endl;

	MBaseline::Convert bconvb(mbref1, mbref0);
	cout << "Back      " << bconv() << endl <<
	  " as " << bconvb(bconv()) << endl;

	Vector<Double> tvec(3);
	tvec = 0.0;
	for (uInt i=MBaseline::ITRF; i<MBaseline::N_Types; i++) {
	  for (uInt j=MBaseline::ITRF; j<MBaseline::N_Types; j++) {
	    MBaseline::Ref rin(i, mf);
	    MBaseline::Ref rout(j, mf);
	    MBaseline mb0(mvb0, rin);
	    MBaseline::Convert forw(rin, rout);
	    MBaseline::Convert backw(rout, rin);
	    if (!allNearAbs(mb0.getValue().getValue().ac() -
			 backw(forw(mb0)).getValue().getValue().ac(), 
			 tvec.ac(), 1e-5)) {
	      cout << MBaseline::showType(i) << " to " <<
		MBaseline::showType(j) << ": " <<
		mb0.getValue().getValue().ac() -
		backw(forw(mb0)).getValue().getValue().ac() << endl;
	    };
	  };
	};
 	{
	  MVDirection mvd0(0.5, 0.5, 0.5);
	  for (uInt i=MDirection::J2000; i<MDirection::N_Types; i++) {
	    for (uInt j=MDirection::J2000; j<MDirection::N_Types; j++) {
	      MDirection::Ref rin(i, mf);
	      MDirection::Ref rout(j, mf);
	      MDirection mb0(mvd0, rin);
	      MDirection::Convert forw(rin, rout);
	      MDirection::Convert backw(rout, rin);
	      if (!allNearAbs(mb0.getValue().getValue().ac() -
			      backw(forw(mb0)).getValue().getValue().ac(), 
			      tvec.ac(), 1e-5)) {
		cout << MDirection::showType(i) << " to " <<
		  MDirection::showType(j) << ": " <<
		  mb0.getValue().getValue().ac() -
		  backw(forw(mb0)).getValue().getValue().ac() << endl;
	      };
	    };
	  };
	}
       
    } catch (AipsError x) {
	cout << x.getMesg() << endl;
    } end_try;

    exit(0);
}
