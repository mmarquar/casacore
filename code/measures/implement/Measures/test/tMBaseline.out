Test measure class MBaseline
--------------------------------------
Baseline: [100, 10, 0]
Baseline: Baseline: [100, 10, 0]
Baseline reference: Reference for an Baseline with Type: ITRF,
Frame: Epoch: 33281::22:09:46.8619
Baseline reference: Reference for an Baseline with Type: J2000
Test Baseline conversion ...
--------------------------------------
Converted Baseline: [100, 10, 0]
 to Reference for an Baseline with Type: J2000
 as Baseline: [100, 10, 0]
